### ISSUE TITLE
EPIC: Improve RPC Equivalency in Rollkit

### ISSUE THREAD

**Issue by nashqueue** [rollkit/rollkit#1268]
### Overview
The current implementation of the Rollkit RPC service does not fully align with the CometBFT rpc specification. This EPIC aims to track all the issues related to improving the RPC equivalency in Rollkit.

### Background
Rollkit RPC provides a set of endpoints for interacting with a Rollkit node. It supports various protocols such as URI over HTTP, JSONRPC over HTTP, and JSONRPC over WebSockets. However, some of the test of the RPC calls do not fully test all the fields mentioned in the documentation.

### Goals
1. Improve the test coverage of the RPC calls to include all fields mentioned in the documentation.
2. Ensure the RPC calls return the correct and expected results.
3. Improve the documentation of the RPC calls to provide more details about their functionality and usage.

### Non-Goals
1. Changes to the core functionality of the Rollkit node are not part of this EPIC.
2. Changes to the protocols supported by the Rollkit RPC service are not part of this EPIC.

```[tasklist]
### Tasks
- [ ] #1258
- [ ] #1259
- [ ] #1260
- [ ] #1261
- [ ] https://github.com/rollkit/rollkit/issues/1269
```

## Functions/methods with no code coverage

```[tasklist]
### rpc/client/Client:
- [ ] Create test for `Client.ABCIQuery` and `Client.ABCIQueryWithOptions`
- [ ] Create test for `Client.Subscribe`
- [ ] Create test for `Client.Genesis`
- [ ] Create test for `Client.NetInfo`
- [ ] Create test for `Client.ConsensusParams`
- [ ] Create test for `Client.BlockResults`
- [ ] Create test for `Client.BroadcastEvidence`
```

```[tasklist]
### rpc/json/service:
- [ ] Create test for `rpc/json/service.Status`
- [ ] Create test for `rpc/json/service.NetInfo`
- [ ] Create test for `rpc/json/service.BlockchainInfo`
- [ ] Create test for `rpc/json/service.Genesis`
- [ ] Create test for `rpc/json/service.GenesisChunked`
- [ ] Create test for `rpc/json/service.BlockByHash`
- [ ] Create test for `rpc/json/service.BlockResults`
- [ ] Create test for `rpc/json/service.Commit`
- [ ] Create test for `rpc/json/service.Tx`
- [ ] Create test for `rpc/json/service.BlockSearch`
- [ ] Create test for `rpc/json/service.Validators`
- [ ] Create test for `rpc/json/service.DumpConsensusState`
- [ ] Create test for `rpc/json/service.GetConsensusState`
- [ ] Create test for `rpc/json/service.ConsensusParams`
- [ ] Create test for `rpc/json/service.UnconfirmedTxs`
- [ ] Create test for `rpc/json/service.NumUnconfirmedTxs`
- [ ] Create test for `rpc/json/service.BroadcastTxCommit`
- [ ] Create test for `rpc/json/service.BroadcastTxSync`
- [ ] Create test for `rpc/json/service.BroadcastTxAsync`
- [ ] Create test for `rpc/json/service.ABCIQuery`
- [ ] Create test for `rpc/json/service.BroadcastEvidence`
```

### Completion Criteria
This EPIC can be considered complete when all the related issues are resolved and the RPC service fully aligns with the CometBFT specification.

**Comment by tzdybal** [comment]
There's also a big unit test commented out, that tests Subscribe methods.

**Comment by tzdybal** [comment]
Posting some of my notes here for visibility. First two issues seems to have very low severity. The last one is an absolute blocker. I'll create separate GH issues for them later.

# Issues

## Version
Problem: hermes require Tendermint/CometBFT version to be reported, not rollkit version.

Solution: workaround (or maybe proper solution) is to set `compat_mode = "0.37"` in `~/hermes/config.toml`

## Status endpoint
This is just a warning, but looks bad in the logs and may be confusing. Additionally fixing this can be useful in general.

Problem: Full client always returns that it's catching up (see: https://github.com/rollkit/rollkit/blob/d7359b02715b15b2bf353166fc042584d6c5e8ac/node/full_client.go#L763)

Solution: remove this hardcode. in aggregator mode, we're always "cought up", when following chain we probably should check block times. What about lazy mode?

Workaround: hardcode `catching up` to false.

## WebSockets
Problem: Hermes is not able to deserialize websocket reposnse.

Detail: Hermes expects JSON-RPC message, but rollkit returns "raw" (REST style) message (not wrapped in JSON-RPC).

Solution: fix WS handler to properly encode WS responses.

**Comment by Manav-Aggarwal** [comment]
The issue described in https://github.com/rollkit/rollkit/issues/944 should work after this issue is resolved

**Comment by tzdybal** [comment]
# How to test with hermes - ugly notes

I was testing #1425 using hermes
see also: https://github.com/rollkit/rollkit/issues/1268#issuecomment-1861572871 - compat_mode is required to be able to actually run rollkit chain with hermes

I was following hermes tutorial, and then created custom rollkit chain that used correct ports
command to run rollkit app: `ggmd start --rollkit.aggregator true --rollkit.da_address 127.0.0.1:9090 --rollkit.da_start_height $DA_BLOCK_HEIGHT --rpc.laddr tcp://127.0.0.1:36657 --p2p.laddr "0.0.0.0:36656" --rpc.grpc_laddr 127.0.0.1:27042 --rpc.laddr` tcp://127.0.0.1:27040  (obviously `ggmd` is an app from gm tutorial, but with different name, to avoid name conflict with `gaia manager` used in hermes tutorial)
and this is a part of hermes config related to the chain
```
[[chains]]
id = 'ibc-1'
type = 'CosmosSdk'
rpc_addr = 'http://localhost:27040/'
grpc_addr = 'http://localhost:27042/'
event_source = { mode = 'push', url = 'ws://localhost:27040/websocket', batch_delay = '200ms' }
rpc_timeout = '15s'
trusted_node = true
account_prefix = 'ggm'
key_name = 'wallet'
store_prefix = 'ibc'
gas_price = { price = 0.001, denom = 'stake' }
gas_multiplier = 1.2
default_gas = 1000000
max_gas = 10000000
max_msg_num = 30
max_tx_size = 2097152
clock_drift = '5s'
max_block_time = '30s'
trusting_period = '14days'
trust_threshold = { numerator = '2', denominator = '3' }
compat_mode = "0.37" 
```

before starting hermes, I use: export RUST_LOG=debug to see detailed errors
tested quickly with hermes, and I can't see WS related errors :smile:
new blocks are properly decoded (in debug log of hermes)
but there are also some error messages in rollkit log :confused:

**Comment by gupadhyaya** [comment]
tested all RPC endpoints in https://github.com/rollkit/cosmos-sdk/blob/main/client/docs/swagger-ui/swagger.yaml

Apart from following endpoint, rest all works as expected

```
http://localhost:1317/cosmos/base/tendermint/v1beta1/abci_query
{"code":13,"message":"ABCI Query handler undefined","details":[]}

http://localhost:1317/cosmos/base/node/v1beta1/status
{"code":12,"message":"unknown method Status for service cosmos.base.node.v1beta1.Service","details":[]}

http://localhost:1317/cosmos/mint/v1beta1/annual_provisions
/cosmos/mint/v1beta1/inflation
params
{"code":12,"message":"unknown service cosmos.mint.v1beta1.Query","details":[]}

http://localhost:1317/cosmos/tx/v1beta1/decode
encode
http://localhost:1317/cosmos/tx/v1beta1/simulate
{"code":12,"message":"Not Implemented","details":[]}

http://localhost:1317/cosmos/tx/v1beta1/txs
{"code":3,"message":"must declare at least one event to search","details":[]}

http://localhost:1317/cosmos/upgrade/v1beta1/authority
{"code":12,"message":"unknown service cosmos.upgrade.v1beta1.Query","details":[]}
```

**Comment by gupadhyaya** [comment]
Header methods needs to be implemented in the service and RPC clients.
```
Header(ctx context.Context, height *int64) (*ctypes.ResultHeader, error)
HeaderByHash(ctx context.Context, hash bytes.HexBytes) (*ctypes.ResultHeader, error)
```

**Comment by gupadhyaya** [comment]
This needs to be added to full/light node client impl
```
// UnsubscribeAll unsubscribes given subscriber from all the queries.
UnsubscribeAll(ctx context.Context, subscriber string) error
```
